source/extensions/quic_listeners/quiche/envoy_quic_server_stream.cc

```diff
@@ -14,50 +14,132 @@
-  NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+
+void EnvoyQuicServerStream::encodeHeaders(const Http::HeaderMap& headers, bool end_stream) {
+  ENVOY_STREAM_LOG(debug, "encodeHeaders (end_stream={}) {}.", *this, end_stream, headers);
+  // QUICHE guarantees to take all the headers. This could cause infinite data to
+  // be buffered on headers stream in Google QUIC implementation because
+  // headers stream doesn't have upper bound for its send buffer. But in IETF
+  // QUIC implementation this is safe as headers are sent on data stream which
+  // is bounded by max concurrent streams limited.
+  // Same vulnerability exists in crypto stream which can infinitely buffer data
+  // if handshake implementation goes wrong.
+  // TODO(#8826) Modify QUICHE to have an upper bound for header stream send buffer.
+  WriteHeaders(envoyHeadersToSpdyHeaderBlock(headers), end_stream, nullptr);
+  local_end_stream_ = end_stream;
 }
-void EnvoyQuicServerStream::encodeData(Buffer::Instance& /*data*/, bool /*end_stream*/) {
-  NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+
+void EnvoyQuicServerStream::encodeData(Buffer::Instance& data, bool end_stream) {
+  ENVOY_STREAM_LOG(debug, "encodeData (end_stream={}) of {} bytes.", *this, end_stream,
+                   data.length());
+  local_end_stream_ = end_stream;
+  // This is counting not serialized bytes in the send buffer.
+  uint64_t bytes_to_send_old = BufferedDataBytes();
+  // QUIC stream must take all.
+  WriteBodySlices(quic::QuicMemSliceSpan(quic::QuicMemSliceSpanImpl(data)), end_stream);
+  if (data.length() > 0) {
+    // Send buffer didn't take all the data, threshold needs to be adjusted.
+    Reset(quic::QUIC_BAD_APPLICATION_PAYLOAD);
+    return;
+  }
+
+  uint64_t bytes_to_send_new = BufferedDataBytes();
+  ASSERT(bytes_to_send_old <= bytes_to_send_new);
+  maybeCheckWatermark(bytes_to_send_old, bytes_to_send_new, *filterManagerConnection());
 }
-void EnvoyQuicServerStream::encodeTrailers(const Http::HeaderMap& /*trailers*/) {
-  NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+
+void EnvoyQuicServerStream::encodeTrailers(const Http::HeaderMap& trailers) {
+  ASSERT(!local_end_stream_);
+  local_end_stream_ = true;
+  ENVOY_STREAM_LOG(debug, "encodeTrailers: {}.", *this, trailers);
+  WriteTrailers(envoyHeadersToSpdyHeaderBlock(trailers), nullptr);
 }
+
 void EnvoyQuicServerStream::encodeMetadata(const Http::MetadataMapVector& /*metadata_map_vector*/) {
+  // Metadata Frame is not supported in QUIC.
   NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
 }
 
-void EnvoyQuicServerStream::resetStream(Http::StreamResetReason /*reason*/) {
-  NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+void EnvoyQuicServerStream::resetStream(Http::StreamResetReason reason) {
+  // Upper layers expect calling resetStream() to immediately raise reset callbacks.
+  runResetCallbacks(reason);
+  if (local_end_stream_ && !reading_stopped()) {
+    // This is after 200 early response. Reset with QUIC_STREAM_NO_ERROR instead
+    // of propagating original reset reason.
```

Comment:
Can we get a bit more detail here?   It isn't obvious how this relates to 200 early response, or how StopReading() translates into Reset(QUIC_STREAM_NO_ERROR)

Reply:
done